core/arch/riscv64/emit_utils.c

```diff
@@ -679,9 +802,50 @@ fill_with_nops(dr_isa_mode_t isa_mode, byte *addr, size_t size)
     return true;
 }
 
+/* Having only one thread register tp shared between app and DR, we steal a register for
+ * DR's TLS base in the code cache, and store DR's TLS base into an private lib's TLS slot
+ * for accessing in C code. On entering the code cache (fcache_enter):
+ * - grab gen routine's parameter dcontext and put it into REG_DCXT
+ * - check for pending signals
+ * - load DR's TLS base into dr_reg_stolen from privlib's TLS
+ */
 void
 append_fcache_enter_prologue(dcontext_t *dcontext, instrlist_t *ilist, bool absolute)
 {
-    /* FIXME i#3544: Not implemented */
-    ASSERT_NOT_IMPLEMENTED(false);
+    ASSERT_NOT_IMPLEMENTED(!absolute &&
+                           !TEST(SELFPROT_DCONTEXT, dynamo_options.protect_mask));
+
+    instr_t *no_signals = INSTR_CREATE_label(dcontext);
+
+    /* Save callee-saved reg in case we return for a signal. */
+    APP(ilist,
+        XINST_CREATE_move(dcontext, opnd_create_reg(DR_REG_A1),
+                          opnd_create_reg(REG_DCXT)));
```

Comment:
I'm not sure if we need to do this on RISC-V (and also on AArch64) as `REG_DCXT` (`a5` on RISC-V, `x5` on AArch64) is not a callee-saved register. I think we can get rid of this one and the corresponding one below, am I missing something?

Reply:
I raised the question to Derek, his answer is yes. 

Here's his answer:

yes on that return path since it is all code we control only care about not clobbering callee-saved (when app code gets involved we do not assume anything about ABI, plus we interrupt at block boundaries, so usually in DR gencode we do not care about ABI conventions, which is maybe why this one is confusing)